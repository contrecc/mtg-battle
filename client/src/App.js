import React, { Component } from 'react';
import { Container, Row, Col, Card, CardBody, Button } from 'reactstrap';
import ResultsAlert from './components/ResultsAlert';
import CardComponent from './components/CardComponent';
import Header from './components/Header';
import InfoSection from './components/InfoSection';
import getCards from './utils/getCards';
import './App.css';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      card1: {
        image_uris: {
          art_crop: '',
          normal: '',
          current: ''
        },
        name: '',
        power: 0,
        toughness: 0
      },
      card2: {
        image_uris: {
          art_crop: '',
          normal: '',
          current: ''
        },
        name: '',
        power: 0,
        toughness: 0
      },
      loading: false,
      winner: false,
      bothStrong: false,
      bothWeak: false,
      winStreak: 0,
      btnClicked: null,
      roundCompleted: false,
      winningCard: '',
      pickedCard: ''
    };

    this.showFullArt = this.showFullArt.bind(this);
    this.calculateWinner = this.calculateWinner.bind(this);
    this.playAgain = this.playAgain.bind(this);
    this.loadCards = this.loadCards.bind(this);
  }

  loadCards() {
    getCards()
      .then(json =>
        this.setState({
          card1: {
            image_uris: {
              art_crop: json[0].image_uris.art_crop,
              normal: json[0].image_uris.normal,
              current: json[0].image_uris.art_crop
            },
            name: json[0].name,
            power: Number(json[0].power),
            toughness: Number(json[0].toughness)
          },
          card2: {
            image_uris: {
              art_crop: json[1].image_uris.art_crop,
              normal: json[1].image_uris.normal,
              current: json[1].image_uris.art_crop
            },
            name: json[1].name,
            power: Number(json[1].power),
            toughness: Number(json[1].toughness)
          }
        })
      )
      .catch(error => console.log(error));
  }

  componentDidMount() {
    this.setState(function(prevState, props) {
      return { loading: true };
    });
    this.loadCards();
    this.setState(function(prevState, props) {
      return { loading: false };
    });
  }

  showFullArt() {
    const { card1, card2 } = this.state;
    card1.image_uris.current = card1.image_uris.normal;
    card2.image_uris.current = card2.image_uris.normal;
    this.setState({
      card1,
      card2
    });
  }

  calculateWinner(e) {
    e.preventDefault();

    this.showFullArt();

    const { power: power1, toughness: toughness1 } = this.state.card1;
    const { power: power2, toughness: toughness2 } = this.state.card2;
    const canCard1Win = power1 >= toughness2 ? true : false;
    const canCard2Win = power2 >= toughness1 ? true : false;
    const bothCardsWin = canCard1Win && canCard2Win ? true : false;
    const neitherCardWins = !canCard1Win && !canCard2Win ? true : false;
    const onlyCard1Wins = canCard1Win && !canCard2Win ? true : false;
    const onlyCard2Wins = !canCard1Win && canCard2Win ? true : false;
    const pickedCard = e.target.id === 'btnCard1' ? 'card1' : 'card2';

    if (bothCardsWin) {
      this.setState({
        bothStrong: true,
        winStreak: this.state.winStreak + 1,
        winningCard: '',
        winner: false,
        pickedCard,
        roundCompleted: true
      });
    } else if (neitherCardWins) {
      this.setState({
        bothWeak: true,
        winStreak: this.state.winStreak + 1,
        winningCard: '',
        winner: false,
        pickedCard,
        roundCompleted: true
      });
    } else if (onlyCard1Wins && pickedCard === 'card1') {
      this.setState({
        winner: true,
        winStreak: this.state.winStreak + 1,
        winningCard: 'card1',
        pickedCard,
        roundCompleted: true
      });
    } else if (onlyCard1Wins && pickedCard === 'card2') {
      this.setState({
        winner: false,
        winStreak: 0,
        winningCard: 'card1',
        pickedCard,
        roundCompleted: true
      });
    } else if (onlyCard2Wins && pickedCard === 'card2') {
      this.setState({
        winner: true,
        winStreak: this.state.winStreak + 1,
        winningCard: 'card2',
        pickedCard,
        roundCompleted: true
      });
    } else if (onlyCard2Wins && pickedCard === 'card1') {
      this.setState({
        winner: false,
        winStreak: 0,
        winningCard: 'card2',
        pickedCard,
        roundCompleted: true
      });
    }
  }

  playAgain() {
    this.loadCards();
    this.setState({
      bothStrong: false,
      bothWeak: false,
      winner: false,
      winningCard: '',
      loading: false,
      roundCompleted: false,
      pickedCard: ''
    });
  }

  render() {
    const {
      loading,
      card1,
      card2,
      bothStrong,
      bothWeak,
      winner,
      winningCard,
      pickedCard,
      roundCompleted
    } = this.state;
    return (
      <div className="App text-center">
        <Header />
        <Container>
          <Row>
            <Col>
              {loading ? (
                <Card>
                  <CardBody>
                    <p>Loading...</p>
                  </CardBody>
                </Card>
              ) : (
                <CardComponent
                  id="card1"
                  card={card1}
                  {...{ winner, winningCard, pickedCard }}
                />
              )}
              {roundCompleted ? (
                <Button
                className="btn-lg"
                id="btnCard1"
                onClick={this.calculateWinner}
                disabled
              >
                {card1.name}
              </Button>
              ) : (
                <Button
                className="btn-lg"
                id="btnCard1"
                onClick={this.calculateWinner}
              >
                {card1.name}
              </Button>
              )}
            </Col>
            <Col>
              {loading ? (
                <Card>
                  <CardBody>
                    <p>Loading...</p>
                  </CardBody>
                </Card>
              ) : (
                <CardComponent
                  id="card2"
                  card={card2}
                  {...{ winner, winningCard, pickedCard }}
                />
              )}
              {roundCompleted ? (
                <Button
                className="btn-lg"
                id="btnCard2"
                onClick={this.calculateWinner}
                disabled
              >
                {card2.name}
              </Button>
              ) : (
                <Button
                className="btn-lg"
                id="btnCard2"
                onClick={this.calculateWinner}
              >
                {card2.name}
              </Button>
              )}
            </Col>
          </Row>
          <Row style={{ marginTop: '60px' }}>
            <Col>
              {winner || bothStrong || bothWeak || winningCard ? (
                <ResultsAlert
                  bothStrong={bothStrong}
                  bothWeak={bothWeak}
                  winner={winner}
                />
              ) : null}
              <h2 className="text-info" style={{marginBottom: '50px'}}>Win Streak: {this.state.winStreak}</h2>
              <Button className="btn-lg" onClick={this.playAgain}>
                Play Again
              </Button>
              <InfoSection />
            </Col>
          </Row>
        </Container>
      </div>
    );
  }
}

export default App;
