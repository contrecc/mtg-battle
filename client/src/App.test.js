import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';

jest.mock('./utils/getCards');

// it('renders without crashing', () => {
//   const div = document.createElement('div');
//   ReactDOM.render(<App />, div);
//   ReactDOM.unmountComponentAtNode(div);
// });

it('renders properly', () => {
  const wrapper = mount(<App />);
  expect(wrapper).toMatchSnapshot();

  const button = wrapper.find('button');
  expect(button.length).toEqual(3);

  const image = wrapper.find('img');
  expect(image.length).toEqual(2);
});

// it('renders a button with an onClick property', () => {
//   //const spy = sinon.spy();

//   const wrapper = mount(<App />);
//   const button = wrapper.find('button').first();
//   //button.props.onClick = spy;
//   //console.log(button.props());
//   button.simulate('click');
//   expect(button.props.onClick).toHaveBeenCalled();
// });

it('fetches cards and renders them on mount', done => {
  const wrapper = shallow(<App />);

  setTimeout(() => {
    wrapper.update();

    const state = wrapper.instance().state;
    expect(state.card1.name).toEqual('Shoreline Ranger');
    expect(state.card2.name).toEqual('Darigaaz Reincarnated');

    //expect(wrapper.find('CardComponent').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });

  done();
});
