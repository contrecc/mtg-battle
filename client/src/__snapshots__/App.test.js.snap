// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div
  className="App text-center"
>
  <div
    style={
      Object {
        "marginBottom": "40px",
        "marginTop": "20px",
      }
    }
  >
    <h1
      className="text-info"
      style={
        Object {
          "fontSize": "50px",
          "textAlign": "center",
        }
      }
    >
      MTG BATTLE
    </h1>
    <p
      className="lead text-center"
    >
      Guess which 
      <strong>
        Magic: The Gathering
      </strong>
       creature would defeat the other in combat!
    </p>
  </div>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="bg-secondary card"
          style={
            Object {
              "marginBottom": "30px",
              "marginLeft": "auto",
              "marginRight": "auto",
              "marginTop": "0px",
              "maxWidth": "350px",
              "padding": "20px",
            }
          }
        >
          <img
            alt="card image"
            className={null}
            height={null}
            src="data:image/svg+xml;charset=utf-8,%3Csvg xmlns%3D'http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg' width%3D'100%' height%3D'150' viewBox%3D'0 0 100% 150'%2F%3E"
            style={
              Object {
                "background": "#6c757d",
              }
            }
            width="100%"
          />
        </div>
        <button
          className="btn-lg btn btn-secondary"
          disabled={false}
          id="btnCard1"
          onClick={[Function]}
          type="button"
        />
      </div>
      <div
        className="col"
      >
        <div
          className="bg-secondary card"
          style={
            Object {
              "marginBottom": "30px",
              "marginLeft": "auto",
              "marginRight": "auto",
              "marginTop": "0px",
              "maxWidth": "350px",
              "padding": "20px",
            }
          }
        >
          <img
            alt="card image"
            className={null}
            height={null}
            src="data:image/svg+xml;charset=utf-8,%3Csvg xmlns%3D'http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg' width%3D'100%' height%3D'150' viewBox%3D'0 0 100% 150'%2F%3E"
            style={
              Object {
                "background": "#6c757d",
              }
            }
            width="100%"
          />
        </div>
        <button
          className="btn-lg btn btn-secondary"
          disabled={false}
          id="btnCard2"
          onClick={[Function]}
          type="button"
        />
      </div>
    </div>
    <div
      className="row"
      style={
        Object {
          "marginTop": "40px",
        }
      }
    >
      <div
        className="col"
      >
        <div>
          <h2
            className="text-info"
            style={
              Object {
                "marginBottom": "40px",
              }
            }
          >
            Win Streak: 
            0
          </h2>
        </div>
        <button
          className="btn-lg btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Play Again
        </button>
        <div
          style={
            Object {
              "marginTop": "50px",
            }
          }
        >
          <hr />
          <h3>
            Site built by 
            <a
              className="text-info"
              href="https://github.com/contrecc"
            >
              Colin Contreary
            </a>
          </h3>
          <p>
            It uses React, Reactstrap, and Express
          </p>
          <p>
            All card data from 
            <a
              className="text-info"
              href="https://scryfall.com"
            >
              Scryfall
            </a>
          </p>
          <p>
            MTG Battle is unofficial Fan Content permitted under the Fan Content Policy. Not approved/endorsed by Wizards. Portions of the materials used are property of Wizards of the Coast. Â©Wizards of the Coast LLC
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
